File: arithmetic.c
// $Id: arithmetic.c,v 1.2 2007/05/18 19:02:12 sd Exp $

// This file is part of Attila, a multi-targeted threaded interpreter
// Copyright (c) 2007, UCD Dublin. All rights reserved.
//
// Attila is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; either version 2
// of the License, or (at your option) any later version.
//
// Attila is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.

// Arithmetic primitives
//
// Single-precision arithmetic. This includes normal arithmetic operations,
// comparisons, constants and bit-level operations

#include "attila.h"


// ---------- Constants ----------

// Largest signed number available
// PRIMITIVE MAX-INT ( -- n )
void
max_int() {
  CELL m = INT_MAX;
  PUSH_CELL(m);
}


// Largessmallest signed number available
// PRIMITIVE MIN-INT ( -- n )
void
min_int() {
  CELL m = INT_MIN;
  PUSH_CELL(m);
}


// ---------- Arithmetic ----------

// Negate the top number
// PRIMITIVE NEGATE ( n -- -n )
void
negate() {
  CELL n;
  POP_CELL(n);
  n *= -1;
  PUSH_CELL(n);
}


// Make the top number positive
// PRIMITIVE ABS ( n -- n' )
void
absolute() {
  CELL n;
  POP_CELL(n);
  if(n < 0) n *= -1;
  PUSH_CELL(n);
}


// Integer addition
// PRIMITIVE + ( a b -- a+b )
void
add() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = a + b;
  PUSH_CELL(c);
}


// Fast increment
// PRIMITIVE 1+ ( a -- a+1 )
void
inc() {
  CELLPTR a;
  TOPPTR_CELL(a);
  (*a)++;
}


// Fast decrement
// PRIMITIVE 1- ( a -- a=1 )
void
dec() {
  CELLPTR a;
  TOPPTR_CELL(a);
  (*a)--;
}


// Fast multiply by two, for masking
// PRIMITIVE 2* ( a -- 2*a )
void
two_times() {
  CELLPTR a;
  TOPPTR_CELL(a);
  (*a) = (*a) << 1;
}


// Fast divide by two, for masking
// PRIMITIVE 2/ ( a -- a/2 )
void
two_divide() {
  CELLPTR a;
  TOPPTR_CELL(a);
  (*a) = (*a) >> 1;
}


// Increment by a cell
// PRIMITIVE CELL+ ( addr -- addr' )
void
inc_cell() {
  CELL a;
  POP_CELL(a);
  a += CELL_SIZE;
  PUSH_CELL(a);
}


// Integer subtraction
// PRIMITIVE - ( a b -- a-b )
void
subtract() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = a - b;
  PUSH_CELL(c);
}


// Integer multiplication
// PRIMITIVE * ( a b -- a*b )
void
multiply() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = a * b;
  PUSH_CELL(c);
}


// Divide a by b leaving the remainder and quotient
// PRIMITIVE /MOD ( a b -- a%b a/b )
void
div_mod() {
  CELL a, b, q, r;

  POP_CELL(b);   POP_CELL(a);
  q= (CELL) (a / b);
  r = (CELL) (a % b );
  PUSH_CELL(r);   PUSH_CELL(q);
}


// Leave the divisor and remainder on the stack
// PRIMITIVE */MOD ( a b c -- (a*b)%c (a*b)/c )
void
star_div_mod() {
  CELL a, b, c, i, q, r;

  POP_CELL(c);  POP_CELL(b);   POP_CELL(a);
  i = a * b;
  q = i / c;   r = i % c;
  PUSH_CELL(r);   PUSH_CELL(q);
}


// Maximum
// PRIMITIVE MAX ( a b -- max )
void
mmax() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a > b) ? a : b;
  PUSH_CELL(c);
}


// Minimum
// PRIMITIVE MIN ( a b -- min )
void
mmin() {
  CELL a, b, c;

 POP_CELL(b);   POP_CELL(a);
  c = (a > b) ? b : a;
  PUSH_CELL(c);
}


// ---------- Bit-level operators ----------

// Bitwise and
// PRIMITIVE AND ( a b -- a&b )
void
bit_and() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = a & b;
  PUSH_CELL(c);
}


// Bitwise or
// PRIMITIVE OR ( a b -- a|b )
void
bit_or() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = a | b;
  PUSH_CELL(c);
}


// Bitwise exclusive-or
// PRIMITIVE XOR ( a b -- a^b )
void
bit_xor() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = a ^ b;
  PUSH_CELL(c);
}


// Bitwise not
// PRIMITIVE INVERT ( a -- ~a )
void
bit_not() {
  CELL a, b;

  POP_CELL(a);
  b = ~a;
  PUSH_CELL(b);
}


// Bitwise left-shift
// PRIMITIVE LSHIFT ( a n -- a<<n )
void
bit_shift_left() {
  CELL a, n, c;

  POP_CELL(n);   POP_CELL(a);
  c = a << n;
  PUSH_CELL(c);
}


// Bitwise right-shift
// PRIMITIVE RSHIFT ( a n -- a>>n )
void
bit_shift_right() {
  CELL a, n, c;

  POP_CELL(n);   POP_CELL(a);
  c = a >> n;
  PUSH_CELL(c);
}


// ---------- Logical tests ----------

// Test greater-than-or-equal-to
// PRIMITIVE >= ( a b -- f )
void
test_gte() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a >= b);
  PUSH_CELL(c);
}


// Test greater-than
// PRIMITIVE > ( a b -- f )
void
test_gt() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a > b);
  PUSH_CELL(c);
}


// Test less-than-or-equal-to
// PRIMITIVE <= ( a b -- f )
void
test_lte() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a <= b);
  PUSH_CELL(c);
}


// Test less-than
// PRIMITIVE < ( a b -- f )
void
test_lt() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a < b);
  PUSH_CELL(c);
}


// Test equal
// PRIMITIVE = ( a b -- f )
void
test_eq() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a == b);
  PUSH_CELL(c);
}


// Test inequal
// PRIMITIVE <> ( a b -- f )
void
test_neq() {
  CELL a, b, c;

  POP_CELL(b);   POP_CELL(a);
  c = (a != b);
  PUSH_CELL(c);
}


// Test for zero
// PRIMITIVE 0= ( n -- f )
void
test_eqzero() {
  CELL a, f;

  POP_CELL(a);
  f = (a == 0);
  PUSH_CELL(f);
}


// Test for non-zero
// PRIMITIVE 0<> ( n -- f )
void
test_neqzero() {
  CELL a, f;

  POP_CELL(a);
  f = (a != 0);
  PUSH_CELL(f);
}


// Test for less than zero
// PRIMITIVE 0< ( n -- f )
void
test_ltzero() {
  CELL n, f;

  POP_CELL(n);
  f = (n < 0);
  PUSH_CELL(f);
}


// Test for greater than zero
// PRIMITIVE 0> ( n -- f )
void
test_gtzero() {
  CELL n, f;

  POP_CELL(n);
  f = (n > 0);
  PUSH_CELL(f);
}


// Test whether a lies between b and c
// PRIMITIVE WITHIN ( a b c -- f )
void
within() {
  CELL a, b, c, f;

  POP_CELL(c);   POP_CELL(b);   POP_CELL(a);
  f = (a <= b) && (a < c);
  PUSH_CELL(f);
}

File: test/arithmetic.fs
\ $Id$

\ This file is part of Attila, a retargetable threaded interpreter
\ Copyright (c) 2007--2010, Simon Dobson <simon.dobson@computer.org>.
\ All rights reserved.
\
\ Attila is free software; you can redistribute it and/or
\ modify it under the terms of the GNU General Public License
\ as published by the Free Software Foundation; either version 2
\ of the License, or (at your option) any later version.
\
\ Attila is distributed in the hope that it will be useful,
\ but WITHOUT ANY WARRANTY; without even the implied warranty of
\ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
\ GNU General Public License for more details.
\
\ You should have received a copy of the GNU General Public License
\ along with this program; if not, write to the Free Software
\ Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111, USA.

\ Tests of arithmetic operations

TESTCASES" Bit-level operations"

TESTING" Basic assumptions"
: BITS-SET? IF 1 ELSE 0 THEN ;
{ 0         BITS-SET? -> 0 }
{ 1         BITS-SET? -> 1 }
{ 1 NEGATE  BITS-SET? -> 1 }


TESTING" AND, OR, NOT, XOR, INVERT"
{ 0 0 AND -> 0 }
{ 0 1 AND -> 0 }
{ 1 0 AND -> 0 }
{ 1 1 AND -> 1 }

{ 0 0 OR -> 0 }
{ 1 0 OR -> 1 }
{ 0 1 OR -> 1 }
{ 1 1 OR -> 1 }

{ 0 0 XOR -> 0 }
{ 1 0 XOR -> 1 }
{ 0 1 XOR -> 1 }
{ 1 1 XOR -> 0 }

{ TRUE NOT  -> FALSE }
{ FALSE NOT -> TRUE }
{ 1 NOT     -> FALSE }

{ 0 INVERT 1 AND -> 1 }
{ 1 INVERT 1 AND -> 0 }

\ Cells full of zeros and full of ones
0         CONSTANT 0S
0S INVERT CONSTANT 1S

{ 0S INVERT -> 1S }
{ 1S INVERT -> 0S }

{ 0S 0S AND -> 0S }
{ 0S 1S AND -> 0S }
{ 1S 0S AND -> 0S }
{ 1S 1S AND -> 1S }

{ 0S 0S OR -> 0S }
{ 1S 0S OR -> 1S }
{ 0S 1S OR -> 1S }
{ 1S 1S OR -> 1S }

{ 0S 0S XOR -> 0S }
{ 1S 0S XOR -> 1S }
{ 0S 1S XOR -> 1S }
{ 1S 1S XOR -> 0S }


TESTING" Bit-level arithmetic operations"

\ Cell with only its most significant bit set
1S 1 RSHIFT INVERT CONSTANT MSB

{ MSB BITS-SET? -> 1 }

{ 0s 2*       -> 0S }
{ 1 2*        -> 2 }
{ 1000 2*     -> 2000 }
{ 1S 2* 1 XOR -> 1S }
{ MSB 2*      -> 0S }

{ 0s 2/          -> 0S }
{ 1 2/           -> 0 }
{ 2 2/           -> 1 }
{ 2000 2/        -> 1000 }
{ 1S 1 XOR 2/    -> 1S }
{ MSB 2/ MSB AND -> MSB }

{ 1 0 LSHIFT        -> 1 }
{ 1 1 LSHIFT        -> 2 }
{ 1 2 LSHIFT        -> 4 }
{ 1S 1 LSHIFT 1 XOR -> 1S }
{ MSB 1 LSHIFT      -> 0 }

{ 1 0 RSHIFT           -> 1 }
{ 1 1 RSHIFT           -> 0 }
{ 2 1 RSHIFT           -> 1 }
{ 4 2 RSHIFT           -> 1 }
{ MSB 1 RSHIFT MSB AND -> 0 }
{ MSB 1 RSHIFT 2*      -> MSB }


TESTING" Arithmetic comparisons"

\ Maximum and minimum values, assuming 2's complement arithmetic
0 INVERT                 CONSTANT MAX-UINT
0 INVERT 1 RSHIFT        CONSTANT MAX-INT
0 INVERT 1 RSHIFT INVERT CONSTANT MIN-INT
0 INVERT 1 RSHIFT        CONSTANT MID-UINT
0 INVERT 1 RSHIFT INVERT CONSTANT MID-UINT+1

{ 0 0=        -> TRUE }
{ 1 0=        -> FALSE }
{ 2 0=        -> FALSE }
{ -1 0=       -> FALSE }
{ MAX-UINT 0= -> FALSE }
{ MIN-INT 0=  -> FALSE }
{ MAX-INT 0=  -> FALSE }

{ 0 0 =   -> TRUE }
{ 1 1 =   -> TRUE }
{ -1 -1 = -> TRUE }
{ 1 0 =   -> FALSE }
{ -1 0 =  -> FALSE }
{ 0 1 =   -> FALSE }
{ 0 -1 =  -> FALSE }

{ 0 0<       -> FALSE }
{ -1 0<      -> TRUE }
{ MIN-INT 0< -> TRUE }
{ 1 0<       -> FALSE }
{ MAX-INT 0< -> FALSE }

{ 0 0<=  -> TRUE }
{ 1 0<=  -> FALSE }
{ -1 0<= -> TRUE }
{ 0 0>=  -> TRUE }
{ 1 0>=  -> TRUE }
{ -1 0>= -> FALSE }

{ 0 1 <             -> TRUE }
{ 1 2 <             -> TRUE }
{ -1 0 <            -> TRUE }
{ -1 1 <            -> TRUE }
{ MIN-INT 0 <       -> TRUE }
{ MIN-INT MAX-INT < -> TRUE }
{ 0 MAX-INT <       -> TRUE }
{ 0 0 <             -> FALSE }
{ 1 1 <             -> FALSE }
{ 1 0 <             -> FALSE }
{ 2 1 <             -> FALSE }
{ 0 -1 <            -> FALSE }
{ 1 -1 <            -> FALSE }
{ 0 MIN-INT <       -> FALSE }
{ MAX-INT MIN-INT < -> FALSE }
{ MAX-INT 0 <       -> FALSE }

{ 0 1 >             -> FALSE }
{ 1 2 >             -> FALSE }
{ -1 0 >            -> FALSE }
{ -1 1 >            -> FALSE }
{ MIN-INT 0 >       -> FALSE }
{ MIN-INT MAX-INT > -> FALSE }
{ 0 MAX-INT >       -> FALSE }
{ 0 0 >             -> FALSE }
{ 1 1 >             -> FALSE }
{ 1 0 >             -> TRUE }
{ 2 1 >             -> TRUE }
{ 0 -1 >            -> TRUE }
{ 1 -1 >            -> TRUE }
{ 0 MIN-INT >       -> TRUE }
{ MAX-INT MIN-INT > -> TRUE }
{ MAX-INT 0 >       -> TRUE }

\ { 0 1 U<               -> TRUE }
\ { 1 2 U<               -> TRUE }
\ { 0 MID-UINT U<        -> TRUE }
\ { 0 MAX-UINT U<        -> TRUE }
\ { MID-UINT MAX-UINT U< -> TRUE }
\ { 0 0 U<               -> FALSE }
\ { 1 1 U<               -> FALSE }
\ { 1 0 U<               -> FALSE }
\ { 2 1 U<               -> FALSE }
\ { MID-UINT 0 U<        -> FALSE }
\ { MAX-UINT 0 U<        -> FALSE }
\ { MAX-UINT MID-UINT U< -> FALSE }

{ 0 1 MIN             -> 0 }
{ 1 2 MIN             -> 1 }
{ -1 0 MIN            -> -1 }
{ -1 1 MIN            -> -1 }
{ MIN-INT 0 MIN       -> MIN-INT }
{ MIN-INT MAX-INT MIN -> MIN-INT }
{ 0 MAX-INT MIN       -> 0 }
{ 0 0 MIN             -> 0 }
{ 1 1 MIN             -> 1 }
{ 1 0 MIN             -> 0 }
{ 2 1 MIN             -> 1 }
{ 0 -1 MIN            -> -1 }
{ 1 -1 MIN            -> -1 }
{ 0 MIN-INT MIN       -> MIN-INT }
{ MAX-INT MIN-INT MIN -> MIN-INT }
{ MAX-INT 0 MIN       -> 0 }

{ 0 1 MAX             -> 1 }
{ 1 2 MAX             -> 2 }
{ -1 0 MAX            -> 0 }
{ -1 1 MAX            -> 1 }
{ MIN-INT 0 MAX       -> 0 }
{ MIN-INT MAX-INT MAX -> MAX-INT }
{ 0 MAX-INT MAX       -> MAX-INT }
{ 0 0 MAX             -> 0 }
{ 1 1 MAX             -> 1 }
{ 1 0 MAX             -> 1 }
{ 2 1 MAX             -> 2 }
{ 0 -1 MAX            -> 0 }
{ 1 -1 MAX            -> 1 }
{ 0 MIN-INT MAX       -> 0 }
{ MAX-INT MIN-INT MAX -> MAX-INT }
{ MAX-INT 0 MAX       -> MAX-INT }


TESTING" Basic arithmetic operations"

{ 0 5 +        -> 5 }
{ 5 0 +        -> 5 }
{ 0 -5 +       -> -5 }
{ -5 0 +       -> -5 }
{ 1 2 +        -> 3 }
{ 1 -2 +       -> -1 }
{ -1 2 +       -> 1 }
{ -1 -2 +      -> -3 }
{ -1 1 +       -> 0 }
{ MID-UINT 1 + -> MID-UINT+1 }

{ 0 5 -          -> -5 }
{ 5 0 -          -> 5 }
{ 0 -5 -         -> 5 }
{ -5 0 -         -> -5 }
{ 1 2 -          -> -1 }
{ 1 -2 -         -> 3 }
{ -1 2 -         -> -3 }
{ -1 -2 -        -> 1 }
{ 0 1 -          -> -1 }
{ MID-UINT+1 1 - -> MID-UINT }

{ 0 1+        -> 1 }
{ -1 1+       -> 0 }
{ 1 1+        -> 2 }
{ MID-UINT 1+ -> MID-UINT+1 }

{ 2 1-          -> 1 }
{ 1 1-          -> 0 }
{ 0 1-          -> -1 }
{ MID-UINT+1 1- -> MID-UINT }

{ 0 NEGATE  -> 0 }
{ 1 NEGATE  -> -1 }
{ -1 NEGATE -> 1 }
{ 2 NEGATE  -> -2 }
{ -2 NEGATE -> 2 }

{ 0 ABS       -> 0 }
{ 1 ABS       -> 1 }
{ -1 ABS      -> 1 }
{ MIN-INT ABS -> MID-UINT+1 }

{ 0 0 *   -> 0 }
{ 0 1 *   -> 0 }
{ 1 0 *   -> 0 }
{ 1 2 *   -> 2 }
{ 2 1 *   -> 2 }
{ 3 3 *   -> 9 }
{ -3 3 *  -> -9 }
{ 3 -3 *  -> -9 }
{ -3 -3 * -> 9 }

{ MID-UINT+1 1 RSHIFT 2 *               -> MID-UINT+1 }
{ MID-UINT+1 2 RSHIFT 4 *               -> MID-UINT+1 }
{ MID-UINT+1 1 RSHIFT MID-UINT+1 OR 2 * -> MID-UINT+1 }
